### PR TITLE
feat(plugins/techdocs-module-addons-contrib): add TechDocsRedirect plugin

### PR DESCRIPTION
**Pull request by matchaxnb**
Summary of the efforts (will copy to description):
- add TechDocsRedirect add-on. This supports redirects as provided by the [mkdocs-redirects](https://github.com/mkdocs/mkdocs-redirects) plugin through HTML parsing of `<meta>` tags. The current implementation fetches the entire raw HTML page to see if it has an adequate `<meta>` tag. This could be optimized with chunked HTML parsing but this may be premature optimization as the client caches probably already kick in.
- add this add-on to the base template. This way, all new projects may benefit from redirects. This is beneficial because refactoring documents may leave 404 Not Found errors which are very detrimental to projects.

What we are trying to fix:
- 404s in documentation because of refactors
- Duplicate contents as a solution to the previous problem

What we are not trying to fix:
- General support for redirects across the entire app. This would require a central repository of sorts and changes to the core routing. Usability problems may arise.
possible fix to #14644

## Hey, I just made a Pull Request!

Add TechDocsRedirect plug-in and docs.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by awanlin — COMMENTED** [review]
Hi @matchaxnb, this is pretty cool, thanks for the contribution, did a quick scan and left some minor comments. I'm sure the maintainers will have their own.

Also, could you add this to the list of available addons? It's this file: https://github.com/backstage/backstage/blob/master/docs/features/techdocs/addons.md#available-addons

File: plugins/techdocs-module-addons-contrib/package.json [hunk 1]
```diff
@@ -44,6 +44,8 @@
     "@material-ui/lab": "4.0.0-alpha.57",
     "@react-hookz/web": "^15.0.0",
     "git-url-parse": "^13.0.0",
+    "react-router": "6.0.0-beta.0 || ^6.3.0",
```
[the remark this answers is not in the document]
Comment (matchaxnb) [reply plugins/techdocs-module-addons-contrib/package.json]: yes thanks!

**Comment by matchaxnb** [comment]
had to run prettier on one of the template files (App.tsx) as it was angry with it.

**Comment by matchaxnb** [comment]
> Hey @matchaxnb, thanks for the suggestion!
> 
> Looking at the implementation of the existing mkdocs plugin [mkdocs-redirects](https://github.com/mkdocs/mkdocs-redirects), I feel like it could work out of the box. Did you explore it, and if so, did you run in to any issues?
> 
> It inserts this HTML into the "old" page and redirects to the "new" by setting the location.
> 
> ```
> <head>
>     <meta charset="utf-8">
>     <title>Redirecting...</title>
>     <link rel="canonical" href="{url}">
>     <meta name="robots" content="noindex">
>     <script>var anchor=window.location.hash.substr(1);location.href="{url}"+(anchor?"#"+anchor:"")</script>
>     <meta http-equiv="refresh" content="0; url={url}">
> </head>
> ```

Hey @agentbellnorm, so that was my first idea, I tried it for a few days, and this won't work:
1/ this module generates hardcoded meta redirect tags that are not interpreted in the shadow dom
2/ if they were interpreted they would probably not play well with React-Router

That's why this module does it that way.

**Comment by matchaxnb** [comment]
I'm not sur e I'm responsible of the CI fails. Can a maintainer check please?

**Comment by freben** [comment]
Indeed - i retriggered them now.

**Comment by matchaxnb** [comment]
I will update the docs and leave it that way for review.

**Comment by agentbellnorm** [comment]
Hey @matchaxnb, Ok, thanks for the explanation!

Here's some background and my recommendation:

We also have the problem of docs moving, and definitely agree that there is a need for supporting redirects. There are two kinds of moves that we'd like to support:
* Pages moving
  * e.g. a software component changes name
* Docs moving locally inside a Page

The first one, when a software component is renamed, split or joined with another component is the most common one for us. In this case the whole page is removed, and "local" redirects won't help. Since the page is gone, there is nothing to point to the new location.

Although we're not sure we recommend this solution, we maintain a central list of redirects that supports both page level moves, and moves inside a page, e.g:
* `/docs/default/component/react-docs -> /docs/default/component/react`
* `/docs/default/component/great-tutorial/dev-environment -> /docs/default/component/great-tutorial/development-environment`

Of course the downside is that doc creators don't own the redirects or necessarily know that this is possible. But it's simple, complete and client side only.

I think that a solution we bring in to the backstage monorepo should solve both problems mentioned above which is why my recommendation is that you:
* Extend the builder docker image and install the `mkdocs-redirects` plugin
* Create an addon in your backstage instance that looks for the redirect tag in a page, and redirects using react-router if the tag is found.

**Comment by andrewthauer** [comment]
@matchaxnb - Based on your suggestion, it sounds like there could be an official `@backstage/plugin-redirects` plugin to solve this for the community. Possibly with a backend database.

**Comment by matchaxnb** [comment]
> * Extend the builder docker image and install the `mkdocs-redirects` plugin
> * Create an addon in your backstage instance that looks for the redirect tag in a page, and redirects using react-router if the tag is found.

Hi!

Ah! That sort of redirector would have to use the `useRawPage` hook then. I'll see to PoC that.

**Comment by matchaxnb** [comment]
> @matchaxnb - Based on your suggestion, it sounds like there could be an official `@backstage/plugin-redirects` plugin to solve this for the community. Possibly with a backend database.

So, what I offer here is a contribution to `plugins/techdocs-module-addons-contrib/` that allows TechDocs pages to handle redirects from `<meta http-equiv="refresh" />` tags as produced by modules such as [MkDocs Redirects](https://github.com/mkdocs/mkdocs-redirects).

I had initially built a metadata-based implementation but with @agentbellnorm's suggestion I decided to give a go to parsing HTML and redirecting based on that. It works and has less dependencies.

I will change my contribution to mkdocs-techdocs-core accordingly.

**Comment by matchaxnb** [comment]
oh but. we just switch to react-router 6.4.4. This might be very good news for this plug-in as I had to implement a substitute for the { relative } option of navigate(). I will wait until we're fully merged to switch to it and remove my substitute code.

File: plugins/techdocs-module-addons-contrib/src/TechDocsRedirect/utilities.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    const lastSlash = basePath.lastIndexOf('/');
+    const basedir = basePath.substring(0, lastSlash);
+    return `${basedir}/${relativePath}`;
+  }
+  const countOfDotdotslash = relativePath.split('../').length - 1;
+  const pathSplit = basePath.split('/');
+  const finalPath = pathSplit
+    .slice(0, pathSplit.length - countOfDotdotslash)
+    .join('/');
+  const rpWithoutDds = relativePath.replace('../', '');
```
[the remark this answers is not in the document]
Comment (matchaxnb) [reply plugins/techdocs-module-addons-contrib/src/TechDocsRedirect/utilities.ts:73]: good catch, bot, will change this code.
```suggestion
  const rpWithoutDds = relativePath.replaceAll('../', '');
```

File: yarn.lock [hunk 3]
```diff
@@ -12432,10 +12436,10 @@ __metadata:
   languageName: node
   linkType: hard
 
-"@remix-run/router@npm:1.0.3":
-  version: 1.0.3
-  resolution: "@remix-run/router@npm:1.0.3"
-  checksum: 7e535f3c24bf6b0a48aff3de70ee74e5402fe93105aa9bd01f275d5d8f5a9a8a7cadec08d18cf5544d4f24bc9d7a091abc1c9d4ea8efb2893e9077f104a51459
+"@remix-run/router@npm:1.0.4":
+  version: 1.0.4
```
Comment (matchaxnb) [inline yarn.lock:12440]: I am not sure why that is.

**Comment by matchaxnb** [comment]
Somehow the repo seemed to be in a unhappy place. I rebased on master and pushed again to see.

**Comment by matchaxnb** [comment]
Summary of the efforts (will copy to description):
- add TechDocsRedirect add-on. This supports redirects as provided by the [mkdocs-redirects](https://github.com/mkdocs/mkdocs-redirects) plugin through HTML parsing of `<meta>` tags. The current implementation fetches the entire raw HTML page to see if it has an adequate `<meta>` tag. This could be optimized with chunked HTML parsing but this may be premature optimization as the client caches probably already kick in.
- add this add-on to the base template. This way, all new projects may benefit from redirects. This is beneficial because refactoring documents may leave 404 Not Found errors which are very detrimental to projects.

What we are trying to fix:
- 404s in documentation because of refactors
- Duplicate contents as a solution to the previous problem

What we are not trying to fix:
- General support for redirects across the entire app. This would require a central repository of sorts and changes to the core routing. Usability problems may arise.

**Comment by matchaxnb** [comment]
I don't get why these 2 E2E tests fail. I am rebased on master and have not touched these files. If a maintainer may take a look that would be useful.

**Comment by agentbellnorm** [comment]
Hey @matchaxnb, appreciate your efforts!

Since we're not certain that this is the ultimate approach to handle redirects, I think our aim should be to enable people who want to use this addon + the `mkdocs-redirects` plugin to do so, without imposing this approach on all backstage users.

Given this aim, I think the way forward is:
* Letting users opt in to this plugin rather than including it in the base template.
* Letting users opt in to `mkdocs-redirects` rather than adding it to `mkdocs-techdocs-core` by extending the [techdocs-container](https://github.com/backstage/techdocs-container), adding the plugin to the extended image, and then using the extended image when running `techdocs-cli` via the `--docker-image` argument. We can add some instructions/documentation here.
* Calling the addon something more concrete, my suggestion would be `<MkdocsRedirects>`.

**Regarding the implementation:**
Is there a specific reason for parsing the html with a full parser?
-  Is it possible to get the url from the shadow dom? Or do we never mount the redirect html into the page?
-  Instead of adding a dependency to a full html parser, could we achieve the same result with for example a regular expression?

The failing check seems to think there's a voulnerability in the code: https://github.com/backstage/backstage/security/code-scanning/256, On that page there are some recommendations, do you think they are applicable?

**Pull request closed without merging**